File: DSA/VivaswanSingh_590023022/day_99/q2.c
#include <stdio.h>

void dfs(char grid[][5], int rows, int cols, int r, int c)
{
    if (r < 0 || r >= rows || c < 0 || c >= cols)
        return;

    if (grid[r][c] != '1')
        return;

    // Mark land as visited
    grid[r][c] = '0';

    // Down
    dfs(grid, rows, cols, r + 1, c);

    // Up
    dfs(grid, rows, cols, r - 1, c);

    // Right
    dfs(grid, rows, cols, r, c + 1);

    // Left
    dfs(grid, rows, cols, r, c - 1);
}

int countLandmasses(char grid[][5], int rows, int cols)
{
    int count = 0;

    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            if (grid[i][j] == '1')
            {
                count++;

                dfs(grid, rows, cols, i, j);
            }
        }
    }

    return count;
}

int main()
{
    char grid[4][5] = {
        {'1', '1', '0', '0', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '1', '0', '0'},
        {'0', '0', '0', '1', '1'}
    };

    int rows = 4;
    int cols = 5;

    int result = countLandmasses(grid, rows, cols);

    printf("Number of landmasses: %d\n", result);

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day99/Q2.java
import java.util.*;

public class Main {

    static int m, n;
    static char[][] grid;

    // Four possible directions: up, down, left, right
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    static void dfs(int r, int c) {

        // Boundary check
        if (r < 0 || r >= m || c < 0 || c >= n) {
            return;
        }

        // Stop if this cell is water
        if (grid[r][c] != '1') {
            return;
        }

        // Mark land as visited
        grid[r][c] = '0';

        // Visit all 4 adjacent cells
        for (int i = 0; i < 4; i++) {
            dfs(r + dr[i], c + dc[i]);
        }
    }

    public static int countLandmasses() {

        int count = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {

                // Found a new landmass
                if (grid[i][j] == '1') {

                    count++;

                    // Mark the complete landmass as visited
                    dfs(i, j);
                }
            }
        }

        return count;
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        m = sc.nextInt();
        n = sc.nextInt();

        grid = new char[m][n];

        for (int i = 0; i < m; i++) {
            grid[i] = sc.next().toCharArray();
        }

        System.out.println(countLandmasses());

        sc.close();
    }
}